include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td

```diff
@@ -0,0 +1,158 @@
+        "::mlir::VectorType">;
+
+
+// Notice: These predicate definitions assume that the type has been verified to
+//         be a ShapedType
+def AIE2MatMulOperand_4x16xi8  : VectorOfShapeAndType<[4, 16], I8>;
+def AIE2MatMulOperand_4x8xi8   : VectorOfShapeAndType<[4, 8], I8>;
+def AIE2MatMulOperand_4x4xi16  : VectorOfShapeAndType<[4, 4], I16>;
+def AIE2MatMulOperand_4x2xi16  : VectorOfShapeAndType<[4, 2], I16>;
+def AIE2MatMulOperand_2x8xi16  : VectorOfShapeAndType<[2, 8], I16>;
+def AIE2MatMulOperand_4x8xi16  : VectorOfShapeAndType<[4, 8], I16>;
+def AIE2MatMulOperand_2x4xi16  : VectorOfShapeAndType<[2, 4], I16>;
+def AIE2MatMulOperand_4x2xi32  : VectorOfShapeAndType<[4, 2], I32>;
+def AIE2MatMulOperand_4x8xbf16 : VectorOfShapeAndType<[4, 8], BF16>;
+
+def AIE2MatMulOperand_16x8xi4  : VectorOfShapeAndType<[16, 8], I4>;
+def AIE2MatMulOperand_8x8xi8   : VectorOfShapeAndType<[8, 8], I8>;
+def AIE2MatMulOperand_8x4xi8   : VectorOfShapeAndType<[8, 4], I8>;
+def AIE2MatMulOperand_8x4xbf16 : VectorOfShapeAndType<[8, 4], BF16>;
+
+def AIE2MatMulOperand_4x8xi32  : VectorOfShapeAndType<[4, 8], I32>;
+def AIE2MatMulOperand_4x4xi32  : VectorOfShapeAndType<[4, 4], I32>;
+def AIE2MatMulOperand_2x8xi64  : VectorOfShapeAndType<[2, 8], I64>;
+def AIE2MatMulOperand_4x4xi64  : VectorOfShapeAndType<[4, 4], I64>;
+def AIE2MatMulOperand_4x4xf32  : VectorOfShapeAndType<[4, 4], F32>;
+
+def AIE2MatMulLHS :
+  AnyTypeOf<[AIE2MatMulOperand_4x16xi8,
+             AIE2MatMulOperand_4x8xi8,
+             AIE2MatMulOperand_4x4xi16,
+             AIE2MatMulOperand_4x2xi16,
+             AIE2MatMulOperand_2x8xi16,
+             AIE2MatMulOperand_4x8xi16,
+             AIE2MatMulOperand_2x4xi16,
+             AIE2MatMulOperand_4x2xi32,
+             AIE2MatMulOperand_4x8xbf16],
+            "a vector compatible with a lhs operand of matrix-multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
+
+def AIE2MatMulRHS :
+  AnyTypeOf<[AIE2MatMulOperand_16x8xi4,
+             AIE2MatMulOperand_8x8xi8,
+             AIE2MatMulOperand_4x8xi8,
+             AIE2MatMulOperand_2x8xi16,
+             AIE2MatMulOperand_8x4xi8,
+             AIE2MatMulOperand_4x8xi16,
+             AIE2MatMulOperand_4x4xi16,
+             AIE2MatMulOperand_2x4xi16,
+             AIE2MatMulOperand_8x4xbf16],
+            "a vector compatible with a rhs operand of matrix-multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
+
+def AIE2MatMulACC :
+  AnyTypeOf<[AIE2MatMulOperand_4x8xi32,
+             AIE2MatMulOperand_4x4xi32,
+             AIE2MatMulOperand_2x8xi64,
+             AIE2MatMulOperand_4x4xi64,
+             AIE2MatMulOperand_4x4xf32],
+            "a vector compatible with an accumulator of matrix-multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
+
+class ShapeDimsMatch<string lhs, int ld, string rhs, int rd> :
+  CPred<Shape<lhs>.result # "[" # ld # "] == " #
+        Shape<rhs>.result # "[" # rd # "]">;
```

Comment:
I thought this existed upstream but I can't find it - good stuff for upstreaming!

Reply:
Not this one, but you just reminded me that "VectorOfShape" does exist in newer versions of MLIR (we were stuck in an old one for quite some time), I should get rid of that one.